File: backend/app/schemas/book.py
"""
Pydantic schemas for book management.
"""
from datetime import datetime
from pydantic import BaseModel, Field, field_validator

from app.models.book_metadata import CLASS_GRADE_OPTIONS
from app.schemas.base import BaseSchema, PaginatedResponse


class BookMetadataCreate(BaseModel):
    """Schema for book metadata during creation."""
    title: str = Field(..., min_length=1, max_length=500)
    authors: str = Field(..., min_length=1, max_length=500)
    class_grade: str = Field(..., min_length=1, max_length=20)

    @field_validator('class_grade')
    @classmethod
    def validate_class_grade(cls, v: str) -> str:
        if v not in CLASS_GRADE_OPTIONS:
            raise ValueError(f"class_grade must be one of: {', '.join(CLASS_GRADE_OPTIONS)}")
        return v


class BookMetadataUpdate(BaseModel):
    """Schema for updating book metadata."""
    title: str | None = Field(None, min_length=1, max_length=500)
    authors: str | None = Field(None, min_length=1, max_length=500)
    class_grade: str | None = None

    @field_validator('class_grade')
    @classmethod
    def validate_class_grade(cls, v: str | None) -> str | None:
        if v is not None and v not in CLASS_GRADE_OPTIONS:
            raise ValueError(f"class_grade must be one of: {', '.join(CLASS_GRADE_OPTIONS)}")
        return v


class BookResponse(BaseSchema):
    """Schema for book list item response."""
    id: int
    title: str
    authors: str
    class_grade: str
    book_type: str
    thumbnail: str | None
    file_size: int
    created_at: datetime


class BookDetailResponse(BookResponse):
    """Schema for book detail response."""
    file_name: str
    updated_at: datetime


class BookListResponse(PaginatedResponse[BookResponse]):
    """Schema for paginated book list response."""
    pass


class CoverUploadResponse(BaseModel):
    """Schema for cover upload response."""
    thumbnail: str

File: frontend/src/types/book.ts
export interface Book {
  id: number;
  title: string;
  authors: string;
  class_grade: string;
  book_type: 'PDF' | 'EPUB';
  thumbnail: string | null;
  file_size: number;
  file_name?: string;
  created_at: string;
  updated_at?: string;
}

export interface BookDetail extends Book {
  file_name: string;
  updated_at: string;
}

export interface BookMetadataCreate {
  title: string;
  authors: string;
  class_grade: string;
  library_type?: 'school' | 'public';
}

export interface BookMetadataUpdate {
  title?: string;
  authors?: string;
  class_grade?: string;
  library_type?: 'school' | 'public';
}

export interface PaginatedBooks {
  items: Book[];
  total: number;
  page: number;
  pages: number;
}

export const CLASS_GRADE_OPTIONS = [
  'General',
  'Class 1',
  'Class 2',
  'Class 3',
  'Class 4',
  'Class 5',
  'Class 6',
  'Class 7',
  'Class 8',
  'Class 9',
  'Class 10',
  'Class 11',
  'Class 12',
] as const;

export type ClassGrade = typeof CLASS_GRADE_OPTIONS[number];